mesonbuild/modules/gobuild.py

```diff
@@ -0,0 +1,256 @@
+            build.CustomTarget,
+            build.CustomTargetIndex,
+            build.GeneratedList,
+        ),
+    )
+    @typed_kwargs(
+        "gobuild.project",
+        CT_INPUT_KW.evolve(name="sources"),
+        KwargInfo(
+            "include_directories",
+            ContainerTypeInfo(list, (str, build.IncludeDirs)),
+            default=[],
+            listify=True,
+        ),
+        KwargInfo(
+            "link_with",
+            ContainerTypeInfo(list, (build.SharedLibrary, build.StaticLibrary)),
+            default=[],
+            listify=True,
+        ),
+        KwargInfo(
+            "tags",
+            ContainerTypeInfo(list, (str)),
+            default=[],
+            listify=True,
+        ),
+        KwargInfo(
+            "native",
+            bool,
+            default=False,
+        ),
+        KwargInfo(
+            "gomod_tidy",
+            bool,
+            default=True,
+        ),
+    )
+    def project(
+        self,
+        state: ModuleState,
+        args: T.Tuple[
+            str, T.List[T.Union[mesonlib.FileOrString, build.GeneratedTypes]]
+        ],
+        kwargs: ProjectKwargs,
+    ) -> ModuleReturnValue:
+        is_cross_build = state.environment.is_cross_build(
+            mesonlib.MachineChoice.BUILD
+            if kwargs["native"]
+            else mesonlib.MachineChoice.HOST
+        )
+
+        cross_compilers = self.detect_tools(state, is_cross_build)
+
+        proj_name, arg_sources = args
+        all_sources = self.interpreter.source_strings_to_files(
+            list(itertools.chain(arg_sources, kwargs["sources"]))
+        )
+        go_sum_target = build.CustomTarget(
+            f"{proj_name}: go mod tidy",
+            state.subdir,
+            state.subproject,
+            state.environment,
+            [
+                self.go_exe,
+                "-C",
+                "@CURRENT_SOURCE_DIR@",
+                "mod",
+                "tidy" if kwargs["gomod_tidy"] else "verify",
+            ],
+            self.interpreter.source_strings_to_files(["go.mod"]),
+            [f"go-mod-tidy-{proj_name}.log"],
+            capture=True,
+            build_always_stale=True,
+        )
+
+        build_dir_abs = Path(state.environment.get_build_dir())
+        target_dir_abs = build_dir_abs / state.backend.get_target_dir(go_sum_target)
+
+        link_dirs = {
+            "-L" + (build_dir_abs / state.backend.get_target_dir(lib)).as_posix()
+            for lib in kwargs["link_with"]
+        }
+        link_targets = {"-l" + lib.name for lib in kwargs["link_with"]}
+
+        if not is_cross_build:
+            # TODO: always override environment variables CC and CXX for native builds.
+            extra_env = {}
+        else:
+            extra_env = asdict(cross_compilers)
+
+        # TODO: Is there a native API to determine executable extensions?
+        exe_extension: str = ".exe" if "windows" in cross_compilers.GOOS else ""
```

Comment:
Is there a native Meson API to determine the executable extension `*.exe`?

Reply:
Not in a publicly accessible way. There's code in `build.py` (in the `Executable` class), but you'd need to do some refactoring to get at it.

Reply:
Got it. I found the code `self.suffix` in the `Executable` class. Sounds like a significant change to expose the method. I will hand-roll the `exe_suffix` logic myself then.